### ISSUE TITLE
Compress local storage to maximize data storage

### ISSUE THREAD

**Issue by jakcodex** [jakcodex/muledump#217]
Local storage is limited to 10MB and there is presently no way to get around this. Users hitting this limit are unable to save changes to their Muledump.

Compressing local storage can provide up to 5x more data storage space.

**Comment by jakcodex** [comment]
SnappyJS has been fully implemented. It is built as a plugin to a compression layer added to `setuptools.storage.compression`. Additional plugins can be added easily with no need for special code to migrate between formats.

When the first compression layer request is made a self test occurs to test if compression works. If it fails then compression enters a permanently bypassed state.

### Compression: off

**Deflate** will bypass and return the provided input.

**Inflate** will validate if it has received compressed input. If the input is compressed it will return decompressed data. If the input is compressed but not in a recognized format (or it is recognized and the self test failed) it will return empty data. If it is not recognized as compressed data it will return the input. Inflate operates when compression is off to ensure a path back to decompressing all data.

## Compression: on

On almost any error, both methods will return the supplied input. The only exception is when garbage compressed data is detected.

### Deflate

Deflate will compress an input if the following conditions are met: 

1. Self test has passed  
1. The input is at least 1KB  
1. The input is a string

The first two requirements can be bypassed with the proper arguments (shown below).

The output format of a compressed input is: 

```
mc:<format>:<base64_encoded_string>
```

Where format is the compression library id and the base64_encoded_string is the encoded compressed data.

SnappyJS is the first format supported and its id is 0. 

```js
//  force a deflate of a string and the resulting output
setuptools.storage.compression.deflate('testing', null, true)
"mc:0:CiRkR1Z6ZEdsdVp3"
```

### Inflate

Inflate checks that an input matches the proper compressed string format and then also determines the compression format. 

When inflate detects a string to be in a compressed format and an error is encountered, then the input data is considered junk and `undefined` is returned. 

```js
//  decompress the input and the resulting output
setuptools.storage.compression.inflate('mc:0:CiRkR1Z6ZEdsdVp3')
"testing"
```

**Comment by jakcodex** [comment]
Released with v9.4.0.
